flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ImplicitConversionEqualsFunctionITCase.java

```diff
@@ -0,0 +1,228 @@
+                TypeConversionTestBuilder.left(INT(), INT_DATA)
+                        .right(INT(), INT_DATA)
+                        .right(BIGINT(), BIG_INT_DATA)
+                        .right(FLOAT(), FLOAT_DATA)
+                        .right(DOUBLE(), DOUBLE_DATA)
+                        .right(DECIMAL(1, 0), DECIMAL_DATA)
+                        .build(),
+                TypeConversionTestBuilder.left(BIGINT(), BIG_INT_DATA)
+                        .right(BIGINT(), BIG_INT_DATA)
+                        .right(FLOAT(), FLOAT_DATA)
+                        .right(DOUBLE(), DOUBLE_DATA)
+                        .right(DECIMAL(1, 0), DECIMAL_DATA)
+                        .build(),
+                TypeConversionTestBuilder.left(FLOAT(), FLOAT_DATA)
+                        .right(FLOAT(), FLOAT_DATA)
+                        .right(DOUBLE(), DOUBLE_DATA)
+                        .right(DECIMAL(1, 0), DECIMAL_DATA)
+                        .build(),
+                TypeConversionTestBuilder.left(DOUBLE(), DOUBLE_DATA)
+                        .right(DOUBLE(), DOUBLE_DATA)
+                        .right(DECIMAL(1, 0), DECIMAL_DATA)
+                        .build());
+    }
+
+    private static List<TestSpec> implicitConversionBetweenTimeAndString() {
+        return Arrays.asList(
+                TypeConversionTestBuilder.left(DATE(), DATE_DATA)
+                        .right(DATE(), DATE_DATA)
+                        .right(STRING(), STRING_DATA_EQUALS_DATE)
+                        .build(),
+                TypeConversionTestBuilder.left(TIME(), TIME_DATA)
+                        .right(STRING(), STRING_DATA_EQUALS_TIME)
+                        .build(),
+                TypeConversionTestBuilder.left(TIMESTAMP(), TIMESTAMP_DATA)
+                        .right(STRING(), STRING_DATA_EQUALS_TIMESTAMP)
+                        .build());
+    }
+
+    // unsupported temporarily
+    private static List<TestSpec> unsupportedImplicitConversionBetweenNumericAndString() {
+        return Arrays.asList(
+                TypeConversionTestBuilder.left(STRING(), STRING_DATA_EQUALS_NUMERIC)
+                        .right(STRING(), STRING_DATA_EQUALS_NUMERIC)
+                        .fail(TINYINT(), TINY_INT_DATA)
+                        .fail(SMALLINT(), SMALL_INT_DATA)
+                        .fail(INT(), INT_DATA)
+                        .fail(BIGINT(), BIG_INT_DATA)
+                        .fail(FLOAT(), FLOAT_DATA)
+                        .fail(DOUBLE(), DOUBLE_DATA)
+                        .fail(DECIMAL(1, 0), DECIMAL_DATA)
+                        .build());
+    }
+
+    static class TypeConversionTestBuilder {
+        private TestSpec testSpec;
+        private DataType leftType;
+        private final List<Object> columnData = new ArrayList<>();
+        private final List<DataType> columnTypes = new ArrayList<>();
+        private final List<Boolean> isSuccessResults = new ArrayList<>();
```

Comment:
The size of `isSuccessResults` is not same as the size of other two lists. which is not easy to understand. We can store the left data in a separate field, the lists store the right type, data and success flag, build the whole list in  to fill `onFieldsWithData` and `andDataTypes` in `build` method.